File: src/main/java/InstanceCheck.java
import org.jetbrains.annotations.NotNull;

public class InstanceCheck {

    private static void isInfinite(@NotNull Number number) {
        if(number instanceof Integer) {
            System.out.println("Not infinite");
        } else if(number instanceof Double) {
            System.out.println("Infinite? " + ((Double) number).isInfinite());
        } else {
            throw new IllegalArgumentException();
        }
    }

    public static void main(String[] args) {
        isInfinite(1);
        isInfinite(2.0 / 0.0);
    }
}

File: src/main/kotlin/InstanceCheck.kt
fun isInfinite(number: Number) = when(number) {
    is Int -> println("Not infinite")
    is Double -> println("Infinite? ${number.isInfinite()}")
    else -> throw IllegalArgumentException()
}

fun main(args: Array<String>) {
    isInfinite(1)
    isInfinite(2.0 / 0.0)
}